### ISSUE TITLE
Economic check on program-sent messages

### ISSUE THREAD

**Issue by NikVolf** [gear-tech/gear#82]
Those are fuzz-like tests that check invariants like
1. Gas+balance is never lost
2. Gas+balance is never created

Possible flows emulated by the fuzzying framework should include at least the following:
- [x] Simple flow built around submitting programs through `submit_program()` and triggering messages processing (including sending outgoing messages to programs and random addresses and generating traps) by sending a number of `send_message()` extrinsics distributed across a range of blocks;
- [ ] Flow with users replies;
- [ ] Flow with mailbox provessing;
- [ ] Flow with wait list rent collection by external players sending arbitrary inputs.

#### UPD by @SabaunT
- [ ] Gas tree invariants to be fuzzed: https://github.com/gear-tech/gear/issues/82#issuecomment-1142371480

**Comment by breathx** [comment]
> Invariants to be checked in `ValueTree` implementation in `pallet_gas`. In anytime and any state of `pallet_gas::GasTree`:
> 
> 1. all non-external nodes have a parent in `GasTree` storage.
> 2. all non-external nodes have at least one ancestor with value (i.e., `ValueNode::node_with_value` procedure always return `Ok`), however this value can be equal to 0. Also there are no guarantees that this ancestor is a parent.
> 3. for all non-external nodes message with id equal to `parent_node.id` has a `destination` equal to `source` of the message with id, equal to current (child) `node.id`.
> 4. all nodes can't have consumed parent with zero refs (there can't be any nodes like that in storage) between calls to `ValueTree::consume`. Therefore, if node is deleted, it is consumed and has zero refs (and zero value, if it is able to hold value and is of `ValueType::SpecifiedLocal` type).
> 5. if there is an existing consumed node, it has non-zero refs counter and a value >= 0 (between calls to `ValueTree::consume`)
> 6. if a sub-tree of `GasTree` is a tree, where root has `ValueType::External` type, then sub-tree's root is always deleted last.
> 
> Some less important:
> 
> 1. Nodes can become consumed only after `GasTree::consume` call.
> 2. Unspec refs counter for the current node is incremented only after `GasTree::split`, where node with `ValueType::UnspecifiedLocal` type is created.
> 3. Spec refs counter for the current node is incremented only after `GasTree::split_with_value`, where node with `ValueType::SpecifiedLocal` type is created.

I want you to create separate issue with this invariants, related for need in prop-tests implementation.
This issue should be closed once #790 merged.

**Comment by SabaunT** [comment]
> > Invariants to be checked in `ValueTree` implementation in `pallet_gas`. In anytime and any state of `pallet_gas::GasTree`:
> > 
> > 1. all non-external nodes have a parent in `GasTree` storage.
> > 2. all non-external nodes have at least one ancestor with value (i.e., `ValueNode::node_with_value` procedure always return `Ok`), however this value can be equal to 0. Also there are no guarantees that this ancestor is a parent.
> > 3. for all non-external nodes message with id equal to `parent_node.id` has a `destination` equal to `source` of the message with id, equal to current (child) `node.id`.
> > 4. all nodes can't have consumed parent with zero refs (there can't be any nodes like that in storage) between calls to `ValueTree::consume`. Therefore, if node is deleted, it is consumed and has zero refs (and zero value, if it is able to hold value and is of `ValueType::SpecifiedLocal` type).
> > 5. if there is an existing consumed node, it has non-zero refs counter and a value >= 0 (between calls to `ValueTree::consume`)
> > 6. if a sub-tree of `GasTree` is a tree, where root has `ValueType::External` type, then sub-tree's root is always deleted last.
> > 
> > Some less important:
> > 
> > 1. Nodes can become consumed only after `GasTree::consume` call.
> > 2. Unspec refs counter for the current node is incremented only after `GasTree::split`, where node with `ValueType::UnspecifiedLocal` type is created.
> > 3. Spec refs counter for the current node is incremented only after `GasTree::split_with_value`, where node with `ValueType::SpecifiedLocal` type is created.
> 
> I want you to create separate issue with this invariants, related for need in prop-tests implementation. This issue should be partially closed once #790 merged.

I'm using prop-test approach to test the invariants in https://github.com/gear-tech/gear/issues/877.
